docs/design/sql/14-stream-to-stream-join.md

```diff
@@ -0,0 +1,232 @@
+# Stream to stream JOIN
+
+### Table of Contents
+
++ [Background](#background)
+    * [Goals](#goals)
++ [Functional Design](#functional-design)
+    * [Summary of Functionality](#summary-of-functionality)
++ [Technical Design](#technical-design)
+    + [Technical Design](#watermarks)
++ [Testing Criteria](#testing-criteria)
+
+|||
+|---|---|
+|Related Jira|[HZ-986](https://hazelcast.atlassian.net/browse/HZ-986)|
+|Document Status / Completeness|DRAFT|
+|Requirement owner|Sandeep Akhouri|
+|Developer(s)|Sasha Syrotenko|
+|Quality Engineer|Bartlomiej Poplawski|
+|Technical Reviewers|Viliam Durina|
+
+### Background
+
+In streaming data processing, there are several scenarios for joins:
+
+- batch to stream: one source can be batch (e.g. table), and the other source can be streaming data source.
+- stream to stream, both data sources can be streaming (e.g. messaging topics) that need to be joined.
+
+#### Goals
+
+Hazelcast already supports static to stream joins. This work is intended to introduce stream to stream joins. The main
+goals are:
+
+- SQL based / Non-Java friendly: easy to use format also for non-Java developers to use stream to stream joins
+- different join types (`INNER` , `LEFT`/`RIGHT`/`FULL` `OUTER` `JOIN`s.
+
+It should be possible to efficiently manage state for these joins using watermarks or other mechanism(s).
+
+### Functional Design
+
+#### Summary of Functionality
```

Comment:
Will the stream-to-stream join be available via Pipeline API?

Reply:
No, we don't plan to do it. Will update it in design doc.

Reply:
Hmm, I think this would be a wrong decision. It will mean that the Pipeline API is not a superset of SQL, it will inevitably lead to _I want to use stream-to-stream joins but also feature X, currently available in Pipeline API only_.

Until SQL is on par with Pipeline API we should support the features in Pipeline API as well. cc @viliam-durina

Reply:
Anyway, it's SQL-specific stream-to-stream design doc, and we may extend it with Pipeline API.

Reply:
+1 for including stream-to-stream in pipeline API as well. I _think_ it won't be a huge change in Pipeline API when the SQL part is done and some customers (like my previous customer :) ) prefer to have strongly-typed Java code instead of SQL queries (especially when logic is complicated).

Reply:
It's a nice-to-have feature for 5.2. My guess is that we won't make it. Pipeline API is type-safe, but SQL, on the other hand, is available in all client languages and doesn't require bytecode upload.